mage_ai/authentication/providers/okta.py

```diff
@@ -0,0 +1,85 @@
+import urllib.parse
+import uuid
+from typing import Dict, Optional
+
+import aiohttp
+from aiohttp import BasicAuth
+
+from mage_ai.authentication.oauth.constants import OAUTH_PROVIDER_OKTA
+from mage_ai.authentication.providers.oauth import OauthProvider
+from mage_ai.authentication.providers.sso import SsoProvider
+from mage_ai.authentication.providers.utils import get_base_url
+from mage_ai.settings.sso import OKTA_CLIENT_ID, OKTA_CLIENT_SECRET, OKTA_DOMAIN_URL
+
+
+class OktaProvider(SsoProvider, OauthProvider):
+    provider = OAUTH_PROVIDER_OKTA
+
+    def __init__(self):
+        self.hostname = OKTA_DOMAIN_URL
+        if not self.hostname.startswith('https'):
+            self.hostname = f'https://{self.hostname}'
+
+    def get_auth_url_response(self, redirect_uri: str = None, **kwargs) -> Optional[Dict]:
+        if OKTA_CLIENT_ID:
+            print('redirect uri:', redirect_uri)
+            base_url = get_base_url(redirect_uri)
+            redirect_uri_query = dict(
+                provider=self.provider,
+                redirect_uri=redirect_uri,
+            )
+            query = dict(
+                client_id=OKTA_CLIENT_ID,
+                redirect_uri=urllib.parse.quote_plus(
+                    f'{base_url}/oauth',
+                ),
+                response_mode='query',
+                response_type='code',
+                scope='openid email profile',
+                state=uuid.uuid4().hex,
+            )
+            query_strings = []
+            for k, v in query.items():
+                query_strings.append(f'{k}={v}')
+
+            return dict(
+                url=f"{self.hostname}/oauth2/default/v1/authorize?{'&'.join(query_strings)}",
+                redirect_query_params=redirect_uri_query,
+            )
+
+    async def get_access_token_response(self, code: str, **kwargs) -> Dict:
+        base_url = get_base_url(kwargs.get('redirect_uri'))
+        data = dict()
+        async with aiohttp.ClientSession() as session:
+            async with session.post(
+                f'{self.hostname}/oauth2/default/v1/token',
+                headers={
+                    'Content-Type': 'application/x-www-form-urlencoded',
+                },
+                data=dict(
+                    grant_type='authorization_code',
+                    code=code,
+                    redirect_uri=f'{base_url}/oauth',
+                ),
+                auth=BasicAuth(OKTA_CLIENT_ID, OKTA_CLIENT_SECRET),
+                timeout=20,
+            ) as response:
+                data = await response.json()
+
+        return data
+
+    async def get_user_info(self, access_token: str = None, **kwargs) -> Dict:
```

Comment:
If access_token is None, do we need to throw exception?
I don't know whether making request with None token will return a good exception message.

Reply:
I think it would throw a 403 exception, but I can add a check anyway